src/metabase/api/collection.clj

```diff
@@ -37,13 +37,44 @@
    [metabase.util.schema :as su]
    [schema.core :as s]
    [toucan.db :as db]
-   [toucan.hydrate :refer [hydrate]]))
+   [toucan.hydrate :refer [hydrate]]
+   [toucan.models :as models]))
 
 ;;; when alias defined for namespaced keywords is run through kondo macro, ns should be regarded as used
 (comment collection.root/keep-me)
 
 (declare root-collection)
 
+(defn- coll-query
+  "Generate a query to return collections viewable by the current user.
+  If provided, additional and clauses will be added to the final where clause.
+  Note that and clauses should be aliased as c (for collection)."
+  [owner-id & and-clauses]
+  {:with      [[:owner-match {:select [:c.personal_owner_id
+                                       [[:concat "/" :c.id "/%"] :pc_prefix]]
+                              :from   [[:collection :c]]
+                              :where  [:not= :c.personal_owner_id nil]}]
+               [:collection-with-owner {:select    [[:c.id :id]
+                                                    [[:coalesce
+                                                      :c.personal_owner_id
+                                                      :om.personal_owner_id] :owner]]
+                                        :from      [[:collection :c]]
+                                        :left-join [[:owner-match :om]
+                                                    [:like :c.location :om.pc_prefix]]}]
+               [:active-ids {:select [[:cwo.id :id]
+                                      [true :selected]]
+                             :from   [[:collection-with-owner :cwo]]
+                             :where  [:or
+                                      [:= :cwo.owner owner-id]
+                                      [:= :cwo.owner nil]]}]]
+   :select    [:c.*]
+   :from      [[:collection :c]]
+   :left-join [[:active-ids :a] [:= :a.id :c.id]]
+   :where     (if and-clauses
+                (into [:and [:= :a.selected true]] and-clauses)
+                [:= :a.selected true])
```

Comment:
The `[:and x]` form will get optimized out by honeysql, so no need for the `if` here.

```
(require '[honeysql.core :as hsql])
(hsql/format {:select [:*]
              :from [:table]
              :where [:and [:= 1 :id] nil nil]})
;; => ["SELECT * FROM table WHERE (1 = id)"]
```

```suggestion
   :where     (into [:and [:= :a.selected true]] and-clauses)
```